asr-python-adapter-finetune-am-conformer-nemo.ipynb

```diff
@@ -0,0 +1,186 @@
+    "\n",
+    "<div align=\"center\">\n",
+    "  <img src=\"https://mermaid.ink/img/pako:eNptkLFqwzAQhl9F3ORAPDSjA4EUx6RgXEjbycpwWOdG1JaMfEoakrx7ZcfpUKrlxH_fz4d0gcoqggTqxp6qAzoW76k0Ipx1-WI6z3sRxyuRF1GOZ3KisK6d3YG8GFdZ9hRJeLbMDRmqvkRGpDLrTuiUiEWUigBtlyIVqzBnEqZ66I39dcX6iKytKXeUf-wn-286QoFeBMvmu0PTD-EfyXaQpP9JFmP_1XN4S3kfD8W4ue6o18pjc52gYQlzaMm1qFX4msuQSOADtSQhCdfaOupZgjS3QPpOIdNGabYOkhqbnuaAnu3b2VSQsPP0gFKNnw7bibr9AJkZdXU\" height=100% />\n",
+    "</div>\n",
+    "\n",
+    "-----"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Advantages and Limitations of Adapter Training\n",
+    "\n",
+    "Adapters can be used with limited amounts of training data and compute budget, but they impose the restriction that the model's original vocabulary must be used - therefore a new characterset vocabulary / tokenizer cannot be used.\n",
+    "\n",
+    "Please refer to the **Advantages of Adapters** and **Limitations of Adapters** sections in the [NeMo Tutorial](#asr-with-adapters) for further details."
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Preparing the Acoustic Encoder for Adapter Training\n",
+    "\n",
+    "Pre-trained models do not automatically support Adapter modules, so we must prepare the acoustic encoder prior to creating the model.\n",
+    "The steps for this procedure are detailed in the **Prepare the \"base\" model** section of the [NeMo Tutorial](#asr-with-adapters)."
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Preparing the model and dataset for adaptation\n",
+    "\n",
+    "There are very few differences between fine-tuning and adapter training with respect to the preparation of the model, data loaders or optimization method.\n",
+    "\n",
+    "One significant difference is that there is no need to change the vocabulary of the pre-trained network, therefore there is no need to construct a new tokenizer or update the character vocabulary. If your customization dataset contains tokens that requires such changes (for example tokens that dont exist in the model), then you will need to either preprocess the text to remove such tokens, or instead perform fine-tuning (which may require substantially more data).\n",
+    "\n",
+    "Another minor difference is the use of spectrogram augmentation (SpecAugment). During fine-tuning, the amount of data is very little compared to when the model was pre-trained with thousands of hours of speech, so SpecAugment may have been used to prevent overfitting. For adapters, we recommend first disabling spec augmentation and trying to see if results are good enough, then adding it back slowly as needed to see if any improvements can be obtained.\n",
```

Comment:
```suggestion
    "Another minor difference is the use of spectrogram augmentation (SpecAugment). During fine-tuning, the amount of data is very little compared to when the model was pre-trained with thousands of hours of speech, so SpecAugment may have been used to prevent overfitting. For adapters, we recommend first disabling spec augmentation and trying to see if results are good enough, then adding it back slowly, as needed, to see if any improvements can be obtained.\n",
```